app/api/v0_1/demand.py

```diff
@@ -1,8 +1,119 @@
+        demands = data['demands']
+    else:
+        raise errors.InvalidUsage("'demands' missing in request data")
+
+
+    params = ["latitude","longitude", "cluster_id", "unit_name", "quantity"]
+
+    # Checking if each element is valid
+    for demand in demands:
+
+        error = check_demand(demand)
+        if error:
+            return error
+
+        # Filtering the dict for safety
+        demand = {param:demand[param] for param in params}
+
+    demand_entries = []
+
+    # Adding demands to database
+    for demand in demands:
+        unit = Unit.query.filter_by(name=demand['unit_name']).first()
+        demand_entry = Demand(latitude=demand['latitude'],longitude=demand['longitude'],units=demand['quantity'],cluster_id=demand['cluster_id'],unit_id=unit.id)
+        db.session.add(demand_entry)
+        demand_entries.append(demand_entry)
+
+    db.session.commit()
+
+    for demand,demand_entry in zip(demands, demand_entries):
+        demand['id'] = demand_entry.id
+
+    # for unit in data['add_units']:
+    #     print(f"Adding unit '{unit}'")
+    #     un = Unit(name=unit)
+    #     db.session.add(un)
+    #     db.session.commit()
+
+
+    return jsonify(demands)
+
+
+def check_demand(demand:Dict[str,str]):
+    """Return error in demand if any"""
+
+    params = ["latitude","longitude", "cluster_id", "unit_name", "quantity" ]
+
+    # Checking if all input parameters are present and are lists
+    for param in params:
+        if param not in demand:
+            raise errors.InvalidUsage(f"Incorrect demand!", invalid_object=demand)
+
+    # if is_float(demand['quantity']):
+    #     demand['quantity'] = float(demand['quantity'])
+    # else:
+    #     raise errors.InvalidUsage("Invalid quantity", invalid_object=demand)
+
+    if demand['quantity'] < 0:
+        raise errors.InvalidUsage("Invalid quantity", invalid_object=demand)
+
+    # if is_float(demand['latitude']):
+    #     demand['latitude'] = float(demand['latitude'])
+    # else:
+    #     raise errors.InvalidUsage("Invalid latitude", invalid_object=demand)
+
+    if demand["latitude"] < -90 or 90 < demand["latitude"]:
+        raise errors.InvalidUsage("Invalid latitude", invalid_object=demand)
+
+    # if is_float(demand['longitude']):
+    #     demand['longitude'] = float(demand['longitude'])
+    # else:
+    #     raise errors.InvalidUsage("Invalid longitude", invalid_object=demand)
+
+    if demand["longitude"] < -180 or 180 < demand["longitude"]:
+        raise errors.InvalidUsage("Invalid longitude", invalid_object=demand)
+
+    if is_int(demand['cluster_id']):
+        demand['cluster_id'] = int(demand['cluster_id'])
+    else:
+        raise errors.InvalidUsage("Invalid cluster_id, should be int", invalid_object=demand)
+
+    all_units = [unit.name for unit in Unit.query.all()]
+
+    if demand['unit_name'] not in all_units:
+        raise errors.InvalidUsage(f"Unit : {demand['unit']} is invalid.", invalid_object=demand)
```

Comment:
Is this supposed to restrict ```unit``` definitions? I'd like for ```unit```s to be anything to the user.

Reply:
It restricts the units to units in the Unit table. Plan was to have units in the database preconfigured and use them.

Reply:
yea that's what I wanted to avoid. so it would be a user-defined resource. so demand units would be a number (float/int) and then a field ```unit_name``` would be required to define units. only unique units would be serialized. this way a model can be defined using whatever units. someone could have *pallets*, someone could have *orders*, or *bananas*, *peanut butter*, etc. Lol. my goal is a long term one for this. Eventually I'd like to be able to build data analysis around the different data types that drive models. That way I can build some kind of feedback loop for the model development. I could eventually give users context around how certain unit types might impact their model.